src/core/net/tcp6.cpp

```diff
@@ -136,20 +156,130 @@ Error Tcp::Endpoint::Abort(void)
 
 Error Tcp::Endpoint::Deinitialize(void)
 {
-    return kErrorNotImplemented;
+    Error     error;
+    Endpoint *prev;
+
+    Tcp &tcp = GetInstance().Get<Tcp>();
+
+    SuccessOrExit(error = tcp.mEndpoints.Find(*this, prev));
+    tcp.mEndpoints.PopAfter(prev);
+    SetNext(nullptr);
+
+exit:
+    return error;
+}
+
+uint8_t Tcp::Endpoint::TimerFlagToIndex(uint8_t aTimerFlag)
+{
+    OT_UNUSED_VARIABLE(aTimerFlag);
+    /*
+     * TODO: Convert from the timer flag provided by TCPlp to the index in
+     * our timers array.
+     */
+    return 0;
+}
+
+bool Tcp::Endpoint::IsTimerActive(uint8_t aTimerIndex)
+{
+    OT_UNUSED_VARIABLE(aTimerIndex);
+    /* TODO: Check whether TCPlp has marked this timer as active. */
+    return false;
+}
+
+void Tcp::Endpoint::SetTimer(uint8_t aTimerFlag, uint32_t aDelay)
+{
+    TimeMilli now         = TimerMilli::GetNow();
+    TimeMilli newFireTime = now + aDelay;
+    uint8_t   timerIndex  = TimerFlagToIndex(aTimerFlag);
+
+    mTimers[timerIndex] = newFireTime.GetValue();
+    otLogDebgTcp("Endpoint %p set timer %u with delay %u", this, static_cast<unsigned int>(timerIndex),
+                 static_cast<unsigned int>(aDelay));
+
+    GetInstance().Get<Tcp>().ResetTimer(false);
+}
+
+void Tcp::Endpoint::CancelTimer(uint8_t aTimerFlag)
+{
+    uint8_t timerIndex = TimerFlagToIndex(aTimerFlag);
+
+    /*
+     * TCPlp has already cleared the timer flag before calling this, so all
+     * that we have to do is reset the main timer if needed.
+     */
+
+    otLogDebgTcp("Endpoint %p cancelled timer %u", this, static_cast<unsigned int>(timerIndex));
+    if (GetInstance().Get<Tcp>().mTimer.GetFireTime() == TimeMilli(mTimers[timerIndex]))
+    {
+        GetInstance().Get<Tcp>().ResetTimer(false);
+    }
```

Comment:
This works too. I updated the code according to your suggestion.

The trade-off is that `mTimer` will expire more often. In contrast, with the current code, it only expires when a TCP timer expires. For context, TCP will set its retransmit timer every RTT, so now we will have `mTimer` expiring once every RTT, even in the common case where the ACK arrives and TCP resets the timer.  The trade-off is what you consider to be expensive.  In previous platforms I've worked with, waking up the CPU from sleep was expensive, so I wrote the timer code to avoid expiring the timer unless we absolutely had to. For OpenThread it's unclear, so I went with your simpler alternative for now. It looks like the default alarm driver wakes up the CPU periodically anyway so it won't make a difference. But if the driver is ever changed to manage the timer in a tickless way, it might be worth revisiting this.

Reply:
To clarify, this is a trade-off between two types of overhead: spurious wakeups from sleep, and additional mutations to the linked list of timers. The old method, which maintained the invariant the `mTimer` is set to the expiry time of the earliest TCP timer requires fewer wakeups from sleep, but requires more linked list mutations to maintain that invariant.

If later on, we decide that we want to minimize the number of times `mTimer` fires for performance reasons, at the expense of mutating the linked list more frequently by setting and cancelling `mTimer`, the best way to do that is to add a single call to `ResetTimer` after processing each input packet. That way, traversing the linked list of endpoints can happen once per input packet, not once each time TCPlp sets a timer (which happens multiple times per input packet). I've added a comment in the code to this effect, in case we revisit this point earlier.